File: strict.go
package main

import (
	"fmt"
	"io"
	"sort"
	"strings"
)

// widgetAllowedPropKeys lists known props keys per widget type (WidgetRegistryV1).
// Unknown props keys warn in default mode and fail when --strict is set.
var widgetAllowedPropKeys = map[string]map[string]struct{}{
	"intro":         keySet("title", "paragraphs"),
	"cover_banner":  keySet("src", "alt"),
	"follow_us":     keySet("title"),
	"info_grid":     keySet("title", "items"),
	"images_grid":   keySet("title", "images"),
	"careers_tabs":  keySet("title", "vacancies", "labels"),
	"apps_showcase": keySet("section_title", "apps"),
	"project_grid":  keySet("heading", "subheading", "section_id", "cards", "min_card_column_width"),
	"media_swiper":  keySet("images", "aria_label"),
	"row":           keySet("children", "gap"),
	"column":        keySet("children"),
	"grid":          keySet("children", "min_column_width", "gap"),
}

func unknownWidgetPropKeyWarning(pagePath, widgetPath, key string) ConfigWarning {
	return contentWarning(pagePath, fmt.Sprintf("%s.props: unknown props key: %q", widgetPath, key))
}

func unknownWidgetPropKeyWarnings(pagePath string, widgets []WidgetNode) []ConfigWarning {
	var warnings []ConfigWarning

	var walk func(pathPrefix string, nodes []WidgetNode)
	walk = func(pathPrefix string, nodes []WidgetNode) {
		for i, w := range nodes {
			wpath := fmt.Sprintf("%s[%d]", pathPrefix, i)
			widgetType := strings.TrimSpace(w.Type)
			allowed, ok := widgetAllowedPropKeys[widgetType]
			if !ok {
				continue
			}
			for key := range w.Props {
				if _, ok := allowed[key]; !ok {
					warnings = append(warnings, unknownWidgetPropKeyWarning(pagePath, wpath, key))
				}
			}
			switch widgetType {
			case "row", "column", "grid":
				children, err := widgetLayoutChildren(w.Props)
				if err != nil || len(children) == 0 {
					continue
				}
				walk(wpath+".props.children", children)
			}
		}
	}

	walk("widgets", widgets)
	return warnings
}

func isStrictEligibleWarning(w ConfigWarning) bool {
	if w.Key != "" {
		return true
	}
	return strings.Contains(w.Detail, "unknown props key:")
}

func enforceStrictWarnings(warnings []ConfigWarning) error {
	var failures []string
	for _, w := range warnings {
		if isStrictEligibleWarning(w) {
			failures = append(failures, w.String())
		}
	}
	sort.Strings(failures)
	if len(failures) == 0 {
		return nil
	}
	if len(failures) == 1 {
		return fmt.Errorf("%s", failures[0])
	}
	return fmt.Errorf("strict validation failed (%d issues):\n  - %s", len(failures), strings.Join(failures, "\n  - "))
}

func handleConfigWarnings(stderr io.Writer, warnings []ConfigWarning, strict bool) error {
	printConfigWarnings(stderr, warnings)
	if strict {
		return enforceStrictWarnings(warnings)
	}
	return nil
}

File: crates/core/src/strict.rs
use crate::error::{CoreError, CoreResult};
use crate::types::ConfigWarning;
use std::io::Write;

pub fn enforce_strict_warnings(warnings: &[ConfigWarning]) -> CoreResult<()> {
    let mut failures: Vec<String> = warnings
        .iter()
        .filter(|w| is_strict_eligible_warning(w))
        .map(|w| w.to_string())
        .collect();
    failures.sort();
    match failures.len() {
        0 => Ok(()),
        1 => Err(CoreError::msg(failures[0].clone())),
        n => Err(CoreError::msg(format!(
            "strict validation failed ({n} issues):\n  - {}",
            failures.join("\n  - ")
        ))),
    }
}

fn is_strict_eligible_warning(w: &ConfigWarning) -> bool {
    if !w.key.is_empty() {
        return true;
    }
    w.detail.contains("unknown props key:")
}

pub fn print_config_warnings(stderr: &mut dyn Write, warnings: &[ConfigWarning]) -> CoreResult<()> {
    for w in warnings {
        writeln!(stderr, "Warning: {}", w.to_string())?;
    }
    Ok(())
}

pub fn handle_config_warnings(
    stderr: &mut dyn Write,
    warnings: &[ConfigWarning],
    strict: bool,
) -> CoreResult<()> {
    print_config_warnings(stderr, warnings)?;
    if strict {
        enforce_strict_warnings(warnings)?;
    }
    Ok(())
}

use crate::types::WidgetNode;
use std::collections::HashSet;

fn widget_allowed_keys(widget_type: &str) -> Option<HashSet<&'static str>> {
    let keys: &[&str] = match widget_type {
        "intro" => &["title", "paragraphs"],
        "cover_banner" => &["src", "alt"],
        "follow_us" => &["title"],
        "info_grid" => &["title", "items"],
        "images_grid" => &["title", "images"],
        "careers_tabs" => &["title", "vacancies", "labels"],
        "apps_showcase" => &["section_title", "apps"],
        "project_grid" => &[
            "heading",
            "subheading",
            "section_id",
            "cards",
            "min_card_column_width",
        ],
        "media_swiper" => &["images", "aria_label"],
        "row" => &["children", "gap"],
        "column" => &["children"],
        "grid" => &["children", "min_column_width", "gap"],
        _ => return None,
    };
    Some(keys.iter().copied().collect())
}

pub fn unknown_widget_prop_key_warnings(
    page_path: &str,
    widgets: &[WidgetNode],
) -> Vec<ConfigWarning> {
    let mut warnings = Vec::new();
    walk_widget_prop_keys("widgets", page_path, widgets, &mut warnings);
    warnings
}

fn walk_widget_prop_keys(
    prefix: &str,
    page_path: &str,
    widgets: &[WidgetNode],
    warnings: &mut Vec<ConfigWarning>,
) {
    for (i, w) in widgets.iter().enumerate() {
        let wpath = format!("{prefix}[{i}]");
        let widget_type = w.widget_type.trim();
        if let Some(allowed) = widget_allowed_keys(widget_type) {
            for key in w.props.keys() {
                if !allowed.contains(key.as_str()) {
                    warnings.push(ConfigWarning::content(
                        page_path,
                        format!("{wpath}.props: unknown props key: {key:?}"),
                    ));
                }
            }
        }
        if matches!(widget_type, "row" | "column" | "grid") {
            if let Ok(Some(children)) = crate::config::widget_layout_children(&w.props) {
                if !children.is_empty() {
                    walk_widget_prop_keys(
                        &format!("{wpath}.props.children"),
                        page_path,
                        &children,
                        warnings,
                    );
                }
            }
        }
    }
}
